spmd/tensor/api.py

```diff
@@ -12,6 +12,62 @@
 """
 _DEBUG_STRICT = False
 
+# NOTE [Autograd interaction between torch.Tensor]
+#
+# The autograd functions defined below are being used by the public
+# facing APIs (i.e. from_local, local_tensor) to ensure our Tensor
+# works together with torch.Tensor within autograd engine. This
+# allows DistributedTensor to exist on part of the module hierarchy
+# and still able to calculate gradients across the torch.Tensor and
+# DistributedTensor boundary.
+# As an example, it enables backward computation of the following:
+#   Module A (with all torch.Tensor params)
+#       -> Module B (with all DistributedTensor params)
+#           -> Module C (with all torch.Tensor params)
+class ToTorchTensor(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, input: "Tensor"):  # type: ignore
+        ctx.previous_placement = input.placements
+        ctx.previous_device_mesh = input.device_mesh
+        return input._local_tensor
+
+    @staticmethod
+    def backward(ctx, grad_output: torch.Tensor):  # type: ignore
+        previous_placement = ctx.previous_placement
+        previous_device_mesh = ctx.previous_device_mesh
+        dist_grad = Tensor.from_local(
+            grad_output, previous_device_mesh, previous_placement
+        )
+        return dist_grad
+
+
+class FromTorchTensor(torch.autograd.Function):
+    @staticmethod
+    def forward(ctx, input: torch.Tensor, device_mesh, placements):  # type: ignore
+        ctx.previous_placement = placements
+        ctx.previous_device_mesh = device_mesh
+
+        dist_tensor = Tensor(
+            input,
+            device_mesh,
+            placements,
+        )
+        return dist_tensor
+
+    @staticmethod
+    def backward(ctx, grad_output: "Tensor"):  # type: ignore
+        previous_placement = ctx.previous_placement
+        previous_device_mesh = ctx.previous_device_mesh
+
+        # reshard to the placement when creating DistributedTensor
+        # so that the gradient layout matches, and we could return
+        # local gradients directly
+        if grad_output.placements != previous_placement:
+            grad_output = grad_output.redistribute(
+                previous_device_mesh, previous_placement
+            )
+        return grad_output._local_tensor, None, None
```

Comment:
This should be `grad_output.local_tensor()`. Keep in mind that the backward should be differentiable!